doc/adr/0010-reducing-snapshot-size.md

```diff
@@ -0,0 +1,343 @@
+value (`Balance.empty`, `TransactionReference.empty` or `Hash.empty`).
+
+The problem is that every single `GlobalSnapshot` contains all this data for all addresses. This is suboptimal because
+majority of this data never changes between two consecutive snapshots, and it results in increasing-only size of
+snapshots. In the long run this can cause issues with storage, network and memory.
+
+## Decision
+
+Replacing existing `GlobalSnapshotInfo` type with a product type of:
+1. `PartialInfo` - Contains the same fields as the existing `GlobalSnapshotInfo` type and represents overrides to the 
+    `balances`, `lastTransactionRef` and `lastStateChannelSnapshotHashes` that are applied by this snapshot. 
+    Missing values for an address should be treated as lack of overrides (the value from previous snapshot still holds).
+      
+2. `AggregatedInfo` - Similarly to `PartialInfo` this type also contains the same fields as the existing `GlobalSnapshotInfo` 
+    but it has one additional field `since` of type `SnapshotOrdinal` which indicates the aggregation range - from `since` 
+    to the current snapshot (both inclusive). If `since` is equal to the current snapshot ordinal, 
+    the `AggregatedInfo` is the same as `PartialInfo`. Missing values has the same meaning as in `PartialInfo`.  
+
+```json
+{
+  "ordinal": 1500,
+  "info": {
+    "partial": {
+      "balances": {},
+      "lastTransactionRefs": {},
+      "lastSnapshotHashes": {}
+    },
+    "aggregated": {
+      "since": 1001,
+      "balances": {},
+      "lastTransactionRefs": {},
+      "lastSnapshotHashes": {}
+    }
+  }
+}
+```
+
+### Aggregation schemes
+
+One of the advantages of this representation are flexible aggregation schemes.
+
+Simple scheme: 
+* if ordinal is divisible by 1000 do aggregate since 0 (genesis)
+* otherwise do aggregate since last ordinal divisible by 1000
+```mermaid
+graph RL;
+    A(ordinal = 1999, since = 1001)
+    B(ordinal = 1000, since = 0)
+    A-->B;
+```
+
+More complicated scheme:
+* if ordinal is divisible by 1000 do aggregate since 0 (genesis)
+* if ordinal is divisible by 100 do aggregate since last ordinal divisible by 1000
+* if ordinal is divisible by 10 do aggregate since last ordinal divisible by 100
+* otherwise, do aggregate since last ordinal divisible by 10
+```mermaid
+graph RL;
+    A(ordinal = 1999, since = 1991)
+    B(ordinal = 1990, since = 1901)
+    C(ordinal = 1900, since = 1001)
+    D(ordinal = 1000, since = 0)
+    A-->B;
+    B-->C;
+    C-->D;
+```
+
+### Aggregated view resolution
+
+Algorithm for resolving an aggregated view of all data - from the genesis to the latest snapshot:
+1. Variable initialization:
+    * `snapshot := <latest-snapshot>`
+    * `acc := snapshot.info.aggregated`
+2. Loop:
+    * if `snapshot.info.aggregated.since` is equal to `0` return `acc`
+    * otherwise, get the snapshot with ordinal `snapshot.info.aggregated.since` and assign it to `snapshot`
```

Comment:
I support that only if downloading the last full snapshot is required

Reply:
Well if the snapshot you download is a full snapshot you don't need to download anything else. If it's let's call it `PartialSnapshot` then you need to download the full snapshot that this partial one references as well. I would also like to find some better name that reflects the fact that these partial snapshots only build on top of the full snapshot. Maybe they should be called `DiffSnapshots` or `UpdateSnapshots` or something. I think that referencing the last full snapshot also gives us some freedom in changing the logic of full/partial snapshot creation. From the point of view of someone asking for a snapshot the fact that every full snapshot is created every 1000 ordinals is a magic number. But if we just reference the last full snapshot we can create them every 10, 100, 1000 or whatever dynamically changing number we want (it could rely on the network load as well) and the client doesn't need to follow the logic.

Reply:
>when downloading any of those you can then check if you got back what you requested.

So I've made a mistake in this algorithm. It should say get the snapshot with ordinal `since - 1` instead of `since`. So we're not referencing last `full` snapshot here, but the next one, because we're starting to aggregate after the last `full` snapshot.

Reply:
Actually I'd avoid using "diff" name because it will be misleading. That being said, we don't keep the difference between balances, but the partial update of balances. Hence, "partial" name suits the best IMO.

Reply:
I think that the name proposed offline by @marcinwadon `SnapshotInfo` suits the best. Both `SnapshotInfo` and `AggregatedInfo` could be then composed into a product type `GlobalSnapshotInfos` I think (similar to `GlobalSnapshotTips`).